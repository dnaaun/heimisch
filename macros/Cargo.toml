[package]
name = "macros"
version = "0.0.0"
edition = "2021"

[dependencies]
macros_impl = { path = "../macros_impl" }
syn = { version = "2.0.87", default-features = false, features = ["extra-traits", "fold"] }

[lib]
proc-macro = true
