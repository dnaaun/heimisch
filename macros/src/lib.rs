extern crate proc_macro;

use proc_macro::TokenStream;
use syn::{parse_macro_input, LitStr};

#[proc_macro_derive(AvailMerge, attributes(id))]
pub fn derive_avail_merge(input: TokenStream) -> TokenStream {
    // Parse the input tokens into a syntax tree
    macros_impl::derive_avail_merge(input).into()
}

#[proc_macro_derive(TypesafeIdb, attributes(idb))]
pub fn derive_typesafe_idb(input: TokenStream) -> TokenStream {
    // Parse the input tokens into a syntax tree
    macros_impl::derive_typesafe_idb(input)
}

#[proc_macro_derive(Table, attributes(db))]
pub fn derive_table(input: TokenStream) -> TokenStream {
    // Parse the input tokens into a syntax tree
    macros_impl::derive_table(input)
}

#[proc_macro]
pub fn zwang_routes(input: TokenStream) -> TokenStream {
    // Parse the input tokens into a syntax tree
    macros_impl::zwang_routes(input)
}

#[proc_macro]
pub fn zwang_url(input: TokenStream) -> TokenStream {
    let input = parse_macro_input!(input as LitStr);
    macros_impl::zwang_url(input).into()
}

#[proc_macro_attribute]
pub fn leptos_test_setup(_: TokenStream, input: TokenStream) -> TokenStream {
    macros_impl::leptos_test_setup::leptos_test_setup(input)
}
