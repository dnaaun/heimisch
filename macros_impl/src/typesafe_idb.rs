use convert_case::{Case, Casing};
use darling::{FromDeriveInput, FromField};
use proc_macro2::Span;
use quote::{quote, IdentFragment};
use syn::{self, parse_macro_input, DeriveInput, Ident};

#[derive(FromDeriveInput)]
#[darling(attributes(idb), supports(struct_named))]
struct IdbRepositoryOpts {
    ident: syn::Ident,
    data: darling::ast::Data<(), IdbField>,
}

#[derive(FromField)]
#[darling(attributes(idb))]
struct IdbField {
    ident: Option<syn::Ident>,
    id: Option<bool>,
    ty: syn::Type,
    index: Option<bool>,
}

pub fn derive_typesafe_idb(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
    let input = parse_macro_input!(input as DeriveInput);
    let opts = IdbRepositoryOpts::from_derive_input(&input).expect("Invalid input");

    let struct_name = &opts.ident;
    let struct_name_lowercase_pluralized = Ident::new(
        &(struct_name.to_string().to_lowercase() + "s"),
        struct_name.span(),
    );
    let store_marker_name = quote::format_ident!("{}StoreMarker", struct_name);

    let fields: Vec<_> = match opts.data {
        darling::ast::Data::Struct(ref fields) => fields.fields.iter().collect(),
        _ => panic!("IdbRepository can only be derived for structs"),
    };

    let id_field = fields
        .iter()
        .find(|f| f.id.is_some())
        .expect("There must be exactly one `#[idb(id)]` field");

    let id_field_name = id_field.ident.as_ref().expect("Fields must be named");
    let id_field_type = &id_field.ty;

    let index_structs: Vec<_> = fields
        .iter()
        .filter(|f| f.index.is_some())
        .map(|f| {
            let field_name = f.ident.as_ref().unwrap().to_string();
            let field_type = &f.ty;
            let index_name = Ident::new(
                &(field_name.to_case(Case::Pascal) + "Index"),
                field_name.span().unwrap_or(Span::call_site()),
            );

            quote! {
                pub struct #index_name { }

                impl typesafe_idb::IndexSpec for #index_name {
                    type Store = #struct_name;
                    const NAME: &'static str = #field_name;
                    type Type = #field_type;
                }
            }
        })
        .collect();

    let index_adds: Vec<_> = fields
        .iter()
        .filter(|f| f.index.is_some())
        .map(|f| {
            let field_name_str = f.ident.as_ref().unwrap().to_string();
            quote! {
                .add_index(idb::builder::IndexBuilder::new(
                    #field_name_str.into(),
                    idb::KeyPath::Single(#field_name_str.into()),
                ))
            }
        })
        .collect();

    let output = quote! {
        pub struct #store_marker_name {}

        impl typesafe_idb::StoreMarker<#struct_name> for #store_marker_name {}
        impl typesafe_idb::Store for #struct_name {
            const NAME: &'static str = stringify!(#struct_name_lowercase_pluralized);
            type Marker = #store_marker_name;
            type Id = #id_field_type;

            fn id(&self) -> &Self::Id {
                &self.#id_field_name
            }

            fn object_store_builder() -> idb::builder::ObjectStoreBuilder {
                idb::builder::ObjectStoreBuilder::new(Self::NAME)
                    .key_path(Some(idb::KeyPath::new_single(stringify!(#id_field_name))))
                    #(#index_adds)*
            }
        }
        #(#index_structs)*
    };

    output.into()
}
