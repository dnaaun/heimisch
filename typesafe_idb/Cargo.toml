[package]
name = "typesafe_idb"
version = "0.0.0"
edition = "2021"

[dependencies]
idb = "0.6.4"
js-sys = "0.3.72"
serde = { version = "1.0.215", features = ["derive"] }
serde_json = "1.0.133"
tracing = "0.1.40"
wasm-bindgen = "0.2.95"
wasm-bindgen-test = "0.3.45"

[dev-dependencies]
macros = { path = "../macros" }
trybuild = "1.0"

# Otherwise (nightly) rust complains about `[wasm_bindgen_test]`
[lints.rust]
unexpected_cfgs = { level = "warn", check-cfg = ['cfg(wasm_bindgen_unstable_test_coverage)'] }
