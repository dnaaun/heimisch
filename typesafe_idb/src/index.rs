use std::{cell::RefCell, marker::PhantomData};

use idb::Query;
use serde::Serialize;

use crate::{
    serde_abstraction::{from_value, to_value},
    ReactivityTrackers, Store, StoreName,
};

pub trait IndexSpec {
    type Store: Store;
    const NAME: StoreName;

    // The `Eq` requirement is used when doing optimistic updates, and it's not really
    // unrealistic at all to expect things that indexed by indexed db have a `Eq` Rust
    // representation.
    type Type: Serialize + Eq;

    fn get_index_value(row: &Self::Store) -> &Self::Type;
}

pub struct Index<'txn, IS> {
    pub(crate) reactivity_trackers: &'txn RefCell<ReactivityTrackers>,
    pub(crate) actual_index: idb::Index,
    pub(crate) _markers: PhantomData<IS>,
}

impl<IS: IndexSpec> Index<'_, IS> {
    pub async fn get(&self, value: &IS::Type) -> Result<Option<IS::Store>, crate::Error> {
        self.reactivity_trackers
            .borrow_mut()
            .add_bulk_access(IS::Store::NAME);

        Ok(self
            .actual_index
            .get(Query::Key(to_value(value)?))?
            .await?
            .map(from_value)
            .transpose()?)
    }

    pub async fn get_all(&self, value: Option<&IS::Type>) -> Result<Vec<IS::Store>, crate::Error> {
        self.reactivity_trackers
            .borrow_mut()
            .add_bulk_access(IS::Store::NAME);

        Ok(self
            .actual_index
            .get_all(
                value.map(|v| to_value(v).map(Query::Key)).transpose()?,
                None,
            )?
            .await?
            .into_iter()
            .map(from_value)
            .collect::<Result<Vec<_>, _>>()?)
    }
}
