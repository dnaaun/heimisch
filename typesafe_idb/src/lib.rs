#![feature(marker_trait_attr)]
#![feature(type_alias_impl_trait)]

mod chain;
mod db;
mod error;
mod txn;
mod txn_valid_for_store;

mod index;
mod object_store;
pub mod serde_abstraction;
mod store;

pub use db::{TypesafeDb, TypesafeDbBuilder};
pub use error::Error;
pub use index::{Index, IndexSpec};
pub use store::Store;
pub use txn::{Present, ReactivityTrackers, ReadOnly, ReadWrite, Txn, TxnBuilder, TxnMode};
pub use txn_valid_for_store::StoreMarker;
pub use chain::Chain;
