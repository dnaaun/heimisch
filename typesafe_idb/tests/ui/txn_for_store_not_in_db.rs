use macros::TypesafeIdb;
use serde::{Deserialize, Serialize};
use typesafe_idb::{Store, TypesafeDb};

#[derive(Serialize, Deserialize)]
struct InstallationId(u64);

#[derive(Serialize, Deserialize)]
struct RepositoryId(u64);

#[derive(Serialize, Deserialize)]
struct UserId(u64);

#[derive(TypesafeIdb, Serialize, Deserialize)]
struct Repository {
    #[idb(id)]
    id: RepositoryId,

    name: String,

    #[idb(index)]
    installation_id: InstallationId,
}

#[derive(TypesafeIdb, Serialize, Deserialize)]
struct User {
    #[idb(id)]
    id: UserId,

    name: String,

    #[idb(index)]
    repository_id: RepositoryId,
}

pub fn main() {
    async {
        let db = TypesafeDb::builder("just test".into())
            .with_store::<Repository>()
            .build()
            .await
            .unwrap();

        // Should raise error
        User::txn(&db).rw();
    };
}
