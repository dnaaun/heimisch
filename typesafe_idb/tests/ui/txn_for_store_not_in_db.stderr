error[E0277]: the trait bound `(RepositoryStoreMarker, ()): StoreMarker<User>` is not satisfied
  --> tests/ui/txn_for_store_not_in_db.rs:43:18
   |
43 |         db.txn().with_store::<User>();
   |                  ^^^^^^^^^^ the trait `StoreMarker<User>` is not implemented for `(RepositoryStoreMarker, ())`
   |
   = help: the trait `StoreMarker<S>` is implemented for `(Head, Tail)`
note: required by a bound in `TxnBuilder::<'db, DbTableMarkers, TxnTableMarkers, Mode>::with_store`
  --> src/txn.rs
   |
   |     pub fn with_store<H2>(
   |            ---------- required by a bound in this associated function
...
   |         DbTableMarkers: StoreMarker<H2>,
   |                         ^^^^^^^^^^^^^^^ required by this bound in `TxnBuilder::<'db, DbTableMarkers, TxnTableMarkers, Mode>::with_store`
